### PR TITLE
ProjDataInMemory forward iterator

### PR DESCRIPTION
**Pull request by rijobro**
it would be great to have this one too. Gets the buffer, which can be used e.g., in SIRF for dot, norm, etc.

### CODE REVIEW

**Comment by KrisThielemans** [comment]
I can see the attraction, but am wondering about details. We've put in a lot of effort to make STIR access to internals safer, and now we're giving it away.

Should we really give access to the `boost::shared_array<float>`? It ties us down to `boost` (C++17 has `std::shared_ptr<float []>`. We could instead use `float *`. Of course, that implies difficulties with ownership and object life-time. It doesn't seem unreasonable though to say that the underlying object needs to remain alive.

Once we do that, we'd need to guarantee this. A (I believe) safe way to do that is to use the same mechanism as `VectorWithOffset::get_data_ptr` (and `get_const_data_ptr`). Doc is https://github.com/UCL/STIR/blob/8cd126e061fd41d65fb30a006b54b3e6bc897eb4/src/include/stir/VectorWithOffset.inl#L570-L581
If we choose that, the simplest way would be to just use `VectorWithOffset` for the buffer.

A cleaner alternative is to provide iterator access. That way, we hide implementation details, and life-time guarantees will be what people are used to. Sadly, implementing your own iterators isn't as easy as you'd like, see e.g. https://stackoverflow.com/questions/8054273/how-to-implement-an-stl-style-iterator-and-avoid-common-pitfalls. However, in this case, we can just "forward" some typedefs such as
```c++
class ProjDataInMemory
{
private:
  typedef buffer_type boost::shared_array<float>;
public:
  typedef value_type buffer_type::value_type;
  ...
  typedef iterator buffer_type::iterator;
  typedef const_iterator buffer_type::const_iterator;
 
  iterator begin() { return buffer.begin() };
  ...

};
```
This is a pain of course (also C++17 deprecated some `std::iterator` stuff). One way around this would be to use the Boost iterator library (which is what I did in [VectorWithOffset](https://github.com/UCL/STIR/blob/8cd126e061fd41d65fb30a006b54b3e6bc897eb4/src/include/stir/VectorWithOffset.h#L47-L53)). They have an interesting [shared_container](https://www.boost.org/doc/libs/1_73_0/libs/iterator/doc/html/iterator/specialized/shared_container.html) adapter that guarantees lifetime.

All that is still a pain. We could also decide to do the following
```c++
class ProjDataInMemory: public Array<1,float>
```
So now we have iterators, `+=` etc for free. However, it would mean intercepting `grow`, `recycle` and probably a few other things.

not what you signed up for?

**Comment by rijobro** [comment]
I suspected you'd say something along these lines! Obviously I'd much rather have `ProjData` iterators, but doubt I'll have time to implement them.

In the short term, would you accept a `get_data` method, which as you say, returns `float*` and `const float*`. I think it's reasonable to expect the lifetime to match that of the object itself. Plus I'd put it in the doxygen.

**Comment by rijobro** [comment]
By the way, travis will obviously fail until #557 is merged.

**Comment by KrisThielemans** [comment]
Would have to follow same naming as `VectorWithOffset` please.

I see you writing a lot of code that we have though (although you might be writing it in SIRF).

Suggestion: make `buffer` an `Array<1,float>`, add forwarding functions for `get_data_ptr` etc. I strongly suggest to also add forwarding for `operator+=` and whatever (a few lines of code). Add similar forwarding in SIRF and done. If you add forwarding of the typedefs and `begin()` and `end(), you have (good enough) iterators.

**Comment by KrisThielemans** [comment]
#561converts `buffer` to `Array<1,float>`

**Comment by rijobro** [comment]
Will wait for #561 before updating this one.

**Comment by rijobro** [comment]
@KrisThielemans updated to use #561.

**Comment by KrisThielemans** [comment]
pre-C++11 failure. can't use `auto` yet... we need to release 4.1, and move on. sorry

File: src/include/stir/ProjDataInMemory.h [hunk 1]
```diff
@@ -92,20 +95,35 @@ class ProjDataInMemory : public ProjDataFromStream
   virtual void axpby(const float a, const ProjData& x,
                      const float b, const ProjData& y);
 
+  //! start value for iterating through all elements in the array, see full_iterator
+  inline DataType::full_iterator begin_all()
```
Comment (KrisThielemans) [inline src/include/stir/ProjDataInMemory.h:99]: no need to expose `begin_all` etc just stick to "normal" iterators I believe? Up to you, but would need both

Comment (rijobro) [reply src/include/stir/ProjDataInMemory.h]: So just `begin()` and `end()` with `const` and non-`const` versions? (That's what I've done in the most recent commit anyway.)

Comment (KrisThielemans) [reply src/include/stir/ProjDataInMemory.h]: I think having only `begin()` is fine, but having `begin_all()` as well is nicer :smile:

Comment (rijobro) [reply src/include/stir/ProjDataInMemory.h]: ok, added too.

**Comment by KrisThielemans** [comment]
want to kill some travis jobs? I can do the Appveyor ones as well as those take ages

**Comment by KrisThielemans** [comment]
@rijobro, sorry, I'm a bit confused by the various commits, so thought best to say it here. We definitely need `begin()` etc.

**Comment by rijobro** [comment]
I kept `begin` (`const` and not), and added `begin_all` (again `const` and not). And the same for `end`. Total of 8 methods. That what you wanted?

**Comment by KrisThielemans** [comment]
yup. thanks. just couldn't see it in the review, but I guess it was outdated.

File: src/test/test_proj_data_maths.cxx [hunk 2]
```diff
@@ -119,10 +119,23 @@ run_tests()
...
+    pd2.ProjData::axpby(a,x2,b,y2);
 
     check_proj_data_are_equal_and_non_zero(pd1,pd2);
+
+    // Check using iterators
+    ProjDataInMemory pd3(pd1);
+    pd3.fill(0.f);
+    ProjDataInMemory::full_iterator pd_iter = pd3.begin();
+    ProjDataInMemory::const_full_iterator x_iter = x1.begin();
+    ProjDataInMemory::const_full_iterator y_iter = y1.begin();
```
Comment (KrisThielemans) [inline src/test/test_proj_data_maths.cxx:134]: confusing to use `full_iterator` here, even though they're identical

File: src/include/stir/ProjDataInMemory.h [hunk 3]
```diff
@@ -92,20 +92,45 @@ class ProjDataInMemory : public ProjDataFromStream
   virtual void axpby(const float a, const ProjData& x,
                      const float b, const ProjData& y);
 
-    
+  //! typedefs to simplify iterator
+  typedef Array<1,float>::full_iterator full_iterator;
+  typedef Array<1,float>::const_full_iterator const_full_iterator;
```
Comment (KrisThielemans) [inline src/include/stir/ProjDataInMemory.h:97]: doxygen comment only applies for one line, unless you use groups. See e.g. https://github.com/UCL/STIR/blob/fd32b1f516cd7739c4cf292ddfe16144356643e7/src/include/stir/Array.h#L102-L107

Use the opportunity to include typedefs for `iterator` and `const_iterator`.

**Comment by KrisThielemans** [comment]
thanks!

can you do a quick update of release_4.1.htm? We forgot to put things in there, mention axpby for Array, ProjDataInMemory and the iterators. ci skip of course.

**Pull request merged**